### DISCUSSION TITLE
runtime.basename配置能从url里自动识别吗？

### DISCUSSION THREAD

**Discussion by ejfkdev** [web-infra-dev/modern.js#6025] [Q&A] [answered]
ssg静态单页，托管到云端部署，访问路径前缀是跟随部署项目的hash变化的，不方便指定固定前缀的basename
有没有办法指定加载的js、css资源为相对路径？

<img width="1098" alt="image" src="https://github.com/user-attachments/assets/7c9b50a9-c13f-4f29-921e-938cd8d9df8b">

<img width="1134" alt="image" src="https://github.com/user-attachments/assets/18e4ce24-eed5-4acd-b8ee-70e7a59aa531">

**Comment by ejfkdev** [comment]
还有启用了bff有后端api的情况，也需要可以自动适应前缀，不知道有没有办法实现

**Reply by zllkjc** [reply]
你构建的时候无法明确当时的 basename 吗？BFF 无法动态的去做的，除非你的部署容器能传入环境变量。

**Reply by ejfkdev** [reply]
另一个用到bff的插件还没走到云端部署的流程，node项目可能是整个部署到容器里的，托管插件的这个url入口路由到容器，如果是这样的话，后台api可以监听到根目录，前端调用api的时候需要走相对路径，前端能单独配置api前缀也可以

**Comment by ejfkdev** [comment, marked as answer]
解决了

1. 静态资源改相对路径
修改assetPrefix为相对路径
```
  output: {
    assetPrefix: './',
  }
```

2. 使用自控式路由
如果使用内置的路由，在非根目录下部署又不修改basename，页面会无限刷新
代码放到src/App.tsx中

3. 重命名入口，build出来就是index.html了
```
  source: {
    entries: {
      index: {
        entry: './src/App.tsx',
      },
    },
    disableDefaultEntries: true,
  }
```

完整配置：
```typescript
import { appTools, defineConfig } from '@modern-js/app-tools';
import { tailwindcssPlugin } from '@modern-js/plugin-tailwindcss';
import { ssgPlugin } from '@modern-js/plugin-ssg';

// https://modernjs.dev/en/configure/app/usage
export default defineConfig({
  source: {
    entries: {
      index: {
        entry: './src/App.tsx',
      },
    },
    disableDefaultEntries: true,
  },
  html: {
    disableHtmlFolder: true,
  },
  runtime: {
    router: false,
  },
  output: {
    distPath: {
      html: '',
    },
    assetPrefix: './',
    ssg: true,
  },
  plugins: [
    appTools({
      bundler: 'experimental-rspack',
    }),
    tailwindcssPlugin(),
    ssgPlugin(),
  ],
});
```
